### ISSUE TITLE
Bidirectional text support

### ISSUE THREAD

**Issue by MohamedA-Ibrahim** [QuestPDF/QuestPDF#445]
**Describe the bug**
When a text flow direction is set to DirectionFromRightToLeft, the numbers are shown reversed. this can be solved by changing the direction to DirectionFromLeftToRight. but when the text contains both Arabic letters and numbers, the previous workaround won't be suitable as the Arabic text won't be displayed correctly.

**To Reproduce**
1. Write Arabic text with numbers in any container text like "نص عربي 50"
2. Generate the pdf
3. The result will be "نص عربي 05"

**Expected behavior**
Numbers should have the correct direction when used with RTL letters.

**Remarks **
This issue also applies to any text that contains both RTL and LTR letters like Arabic and English.

**Screenshots**
Problem shown in the second column:
![image](https://user-images.githubusercontent.com/39552203/210274259-ea8fc58e-4310-4dc7-930d-8efdce811cab.png)

Numbers are displayed correctly If I change to DirectionFromLeftToRight but that causes the Arabic letters to be reversed: 
![image](https://user-images.githubusercontent.com/39552203/210274269-59239bc4-fadc-4000-b60b-3b5cc8f89530.png)

**Comment by leaders-mindes** [comment]
Hi, I would like to add on the above issue that it's not related to numbers only, the case is also valid for any Arabic (RTL) text contains some English (LTR) word(s) or vise versa (English text contains Arabic words)

Example:
Arabic text (RTL) contains English words, content direction is set to RTL:
![image](https://user-images.githubusercontent.com/112489364/210337611-e72ea5fc-53f4-46f6-9424-601fc637513d.png)

Arabic text (RTL) contains English words, content direction is set to LTR:
![image](https://user-images.githubusercontent.com/112489364/210337879-82f1b090-cf62-4584-ba41-6f9cef046013.png)

The correct text should be:
![image](https://user-images.githubusercontent.com/112489364/210338489-a0f6b195-9050-4796-874c-df978b1d5250.png)

Best regards

**Comment by girlpunk** [comment]
Hi, are you able to give some sample code for this?

Thanks

**Comment by leaders-mindes** [comment]
Yes sure, here is the code used for the above test in my comment:

`var document = Document.Create(container =>
            {
                container.Page(page =>
                {
                    page.Size(PageSizes.A4);
                    page.Margin(2, Unit.Centimetre);
                    page.PageColor(Colors.White);
                    page.DefaultTextStyle(x => x.FontSize(20).FontFamily("Arial"));

                    page.Header()
                        .AlignRight()
                        .Text("بسم الله الرحمن الرحيم")
                        .DirectionFromRightToLeft()
                        .SemiBold().FontSize(36).FontColor(Colors.Blue.Medium);

                    page.Content()
                        .PaddingVertical(1, Unit.Centimetre)
                        .ContentFromRightToLeft()
                        .Column(x =>
                        {
                            x.Spacing(20);

                            x.Item().AlignRight().Text("البرمجة بلغة C# ضمن اطار عمل .Net Core ممتع").DirectionFromRightToLeft();
                            x.Item().Image(Placeholders.Image(200, 100));
                        });

                    page.Footer()
                        .AlignCenter()
                        .ContentFromRightToLeft()
                        .Text(x =>
                        {
                            x.Span("صفحة ");
                            x.CurrentPageNumber();
                        });
                });
            });

            var pdfFile = document.GeneratePdf();`

**Comment by girlpunk** [comment]
Does this achieve the desired output?

```csharp
var document = Document.Create(container =>
{
    container.Page(page =>
    {
        page.Size(PageSizes.A4);
        page.Margin(2, Unit.Centimetre);
        page.PageColor(Colors.White);
        page.DefaultTextStyle(x => x.FontSize(20).FontFamily("Arial"));

        page.Header()
            .AlignRight()
            .Text("بسم الله الرحمن الرحيم")
            .DirectionFromRightToLeft()
            .SemiBold().FontSize(36).FontColor(Colors.Blue.Medium);

        page.Content()
            .PaddingVertical(1, Unit.Centimetre)
            .ContentFromRightToLeft()
            .Column(x =>
            {
                x.Spacing(20);

                x.Item().AlignRight().Text(t => {
                    t.Span("البرمجة بلغة").DirectionFromRightToLeft();
                    t.Span(" C# ");
                    t.Span("ضمن اطار عمل").DirectionFromRightToLeft();
                    t.Span(" .Net Core ");
                    t.Span("ممتع").DirectionFromRightToLeft();
                });
                x.Item().Image(Placeholders.Image(200, 100));
            });

        page.Footer()
            .AlignCenter()
            .ContentFromRightToLeft()
            .Text(x =>
            {
                x.Span("صفحة ");
                x.CurrentPageNumber();
            });
    });
});

var pdfFile = document.GeneratePdf();
```

**Comment by leaders-mindes** [comment]
Yes, it gives the following result:
![image](https://user-images.githubusercontent.com/112489364/210346829-3cbb21a8-1434-4fca-ab55-760835301a87.png)

And this is a correct result and I knew that it'll work like this, but this is not practical in the real cases because the content of a Text or Span or whatever may come from user input on some entity and the user input may contain text with mixed words from two languages (RTL and LTR), so I think QuestPDF should handle it internally to solve the root issue.

BTW really really thank you very much for the efforts on this great library, honestly it's the best and most clean PDF library I knew ever :)

**Comment by girlpunk** [comment]
What you're seeing is expected behaviour in this situation, and can be demonstrated in most applications by putting an RTL override in front of the Latin text. This is functionally equivalent to the `DirectionFromRightToLeft()` call, in that it explicitly specifies the content should be rendered in that direction.

The best workaround I can come up with would be to split the input string based on the codepage of characters in it, as detailed [here](https://stackoverflow.com/a/45628450/1132448), and then render the result as separate spans with or without the direction override as needed.

**Comment by lmingle** [comment]
How about using [BidiSharp](https://github.com/fsufyan/BidiSharp)? It implements the Unicode Bidirectional Algorithm. Feeding the string through that library along with right justifying the output string might provide acceptable formatting.

**Comment by leaders-mindes** [comment]
> It looks as though [MohamedA-Ibrahim](https://github.com/MohamedA-Ibrahim) would like the the [Unicode Bidirectional Algorithm](http://www.unicode.org/reports/tr9/) to be implemented. Maybe name it `DirectionAutomatic()`. Then, if the algorithm works out it, make it the default way of rendering.

I think we both ( me and @MohamedA-Ibrahim ) point to the same issue here, that's why I came here and I didn't open a new issue. 
If it's possible to be implemented in the future I think it'll add too much to the library in the subject of RTL support.

**Comment by MarcinZiabek** [comment]
Indeed, it would be best if the library fully supports such cases. This is tough. First of all, I am a little bit against adding even more dependencies: SkiaSharp is popular and well maintained and still, installing os-specific depenencies is problematic for many people. 

We could either provide a proper instruction in the documentation, even with already written extension method. Or we can try to analyze this code of BidiSharp (or similar libraries) and include it in the QuestPDF directly (following the permissive MIT license). I am very open to discuss this matter :D

Just to show that this problem is extra hard, Spotify was fighting with it too 🤣 https://engineering.atspotify.com/2019/04/right-to-left-the-mirror-world/

**Comment by girlpunk** [comment]
@MarcinZiabek Does QuestPDF handle directional override characters correctly? If so, Imingle's suggestion should work rather nicely (though tbh it doesn't even need something to be overridden for an application to implement it)

**Comment by JimBobSquarePants** [comment]
Does QuestPDF do any text shaping at all? 

Implementing the [Bidi Algorithm](https://github.com/SixLabors/Fonts/blob/570875b2664743067e5edd7ac4b69bb789b6c5b4/src/SixLabors.Fonts/Unicode/BidiAlgorithm.cs) is just one part of the tasks required to correctly render multilingual text. You also have to implement glyph substitution and positioning.

**Comment by MarcinZiabek** [comment]
Yes, text shaping is handled by the HarfBuzzSharp sharp library, maintained by the SkiaSharp team. What is a limiting factor in this case is that the library detects text directionality based on the first character/glyph. Whereas to properly display multilangual text, we need to split it into parts using the BiDi algorithm and then assign proper directionality to every run, instead of relying on HarfBuzzSharp automatic detection.

**Comment by JimBobSquarePants** [comment]
HarfBuzzSharp is detecting based on the first character or QuestPDF is? That a significant bug in the latter if so. 

Feel free to reuse any of the code in the implementation I shared if you want to do something similar. It passes all the tests against Unicode v14.

**Comment by MarcinZiabek** [comment]
> HarfBuzzSharp is detecting based on the first character or QuestPDF is? That a significant bug in the latter if so.

HarfBuzzSharp is performing that detection. By default, QuestPDF is relying on that result, while giving the possibility to override directionality. 

Introducing and improving RTL support is a long running effort. Suffice to say, that everything related to text is just surprisingly complex and difficult. But you surely know it much better than I do 🤣

**Comment by MarcinZiabek** [comment]
> Feel free to reuse any of the code in the implementation I shared if you want to do something similar. It passes all the tests against Unicode v14.

Thank you a lot! 😁

The way of how QuestPDF handles text is not perfect at this moment. The algorithm slowly evolves to cover more cases but it certainly is not valid with Unicode specification. I am not feeling well reusing others work even though the license permits such use. If we decide to reuse your work, I will do my best to contact you and discuss the overall scope. I appreciate how much time you spend on this project.

**Comment by JimBobSquarePants** [comment]
>HarfBuzzSharp is performing that detection. By default, QuestPDF is relying on that result, while giving the possibility to override directionality.

I don't know the HarfbuzzSharp API but that seems to me like a misuse of the method. 

When laying out and shaping text you need to move the codepoints from logical to visual order using the Bidi algorithm before shaping. The codepoints are then grouped by script class and passed to the relevant shaper. 

It's a shame I haven't finished my fonts library yet; it would solve any text layout problems. It's the most difficult thing I've ever worked on and is taking years to complete. I'd recommend using bits as inspiration though as I wouldn't wish anyone going through the same process.

**Comment by MohamedA-Ibrahim** [comment]
Updated the issue as the issue happens when a text contains both RTL and LTR letters as shown by @leaders-mindes

**Comment by MarcinZiabek** [comment]
You may want to use the BidiSharp library to split text into multiple runs, then assign proper direction to each Text.Span. https://github.com/fsufyan/BidiSharp

I am planning to integrate this algorithm into QuestPDF at some point, so this feature will work out of the box.

**Comment by AndreiMaz** [comment]
We also experience this issue in nopCommerce. Please consider a valid fix - https://github.com/nopSolutions/nopCommerce/issues/6614

**Comment by navongol** [comment]
> You may want to use the BidiSharp library to split text into multiple runs, then assign proper direction to each Text.Span. https://github.com/fsufyan/BidiSharp
> 
> I am planning to integrate this algorithm into QuestPDF at some point, so this feature will work out of the box.

any news?

**Comment by MarcinZiabek** [comment]
Yes 😁 This item is currently on the list of the highest long-term priorities. It should be available natively over a couple of months from now. This time my confidence is high, as I have already performed an initial investigation, and I see a promising solution.

Thanks to a recently introduced dual-licensing model, I hope to find funding for feature development. The list includes:
- automated font-subsetter (to reduce PDF file size for any language), 
- bi-directional text support, 
- line-breaking compatible with the Unicode standard,
- text justification.

**Comment by navongol** [comment]
> Yes 😁 This item is currently on the list of the highest long-term priorities. It should be available natively over a couple of months from now. This time my confidence is high, as I have already performed an initial investigation, and I see a promising solution.
> 
> Thanks to a recently introduced dual-licensing model, I hope to find funding for feature development. The list includes:
> 
>     * automated font-subsetter (to reduce PDF file size for any language),
> 
>     * bi-directional text support,
> 
>     * line-breaking compatible with the Unicode standard,
> 
>     * text justification.

Hi,
Is there any progress on this issue?

**Comment by MarcinZiabek** [comment]
> Hi,
> Is there any progress on this issue?

Hi 😀

Yes, progress is being made! 🚀 I'm currently focused on adjusting QuestPDF to depend on a custom version of the Skia library.  This step is crucial as it will help us to achieve multiple objectives at once. It's a complex task, so it might take a bit longer than usual to complete.

For regular updates, you can follow [this discussion ](https://github.com/QuestPDF/QuestPDF/discussions/622)where I'll be sharing the progress.

**Comment by navongol** [comment]
> > Hi,
> > Is there any progress on this issue?
> 
> Hi 😀
> 
> Yes, progress is being made! 🚀 I'm currently focused on adjusting QuestPDF to depend on a custom version of the Skia library. This step is crucial as it will help us to achieve multiple objectives at once. It's a complex task, so it might take a bit longer than usual to complete.
> 
> For regular updates, you can follow [this discussion ](https://github.com/QuestPDF/QuestPDF/discussions/622)where I'll be sharing the progress.

It's been a long time, what's going on?

**Comment by MarcinZiabek** [comment]
The feature is under active development— the alpha version will be released in January 2024 😀

**Comment by rajupaladiya** [comment]
@MarcinZiabek - Do we have any updat on this? Is that fixed and any patch/minor version released / release possible?

**Comment by danirzv** [comment]
This problem also exists with Numerical parts of text in RTL which makes to print "123" as "321"

**Comment by MarcinZiabek** [comment]
I am happy to share that the https://github.com/QuestPDF/QuestPDF/discussions/622 effort will target this issue. The Skia SkParagraph-driven should significantly enhance text rendering capabilities and improve other features.

**Comment by AlexandriaDeveloper** [comment]
> This problem also exists with Numerical parts of text in RTL which makes to print "123" as "321"

yes i am facing same problem and also i can't use direction rtl or ltr cause i am getting mixed data Arabic alphabetic with numbers from database

**Comment by danirzv** [comment]
I'm facing same issue, If there is something I can help with I will be so glad to contribute

**Comment by danirzv** [comment]
> Any update for this? January going to end.... sooonnnn............
it fixed at `2024.3.0-alpha`

**Comment by danirzv** [comment]
@MarcinZiabek thank you for your awesome fast fix which handled previous problem so nice.
there is still a small bug remaining
as I put `rtl` culture data which containing date it orders wrong way
![image](https://github.com/QuestPDF/QuestPDF/assets/14872370/7fcd2e36-9454-44d0-ab28-327796383287)

**Comment by MarcinZiabek** [comment]
@danirzv I am not an RTL expert, so please help me better understand this issue 😄 

I understand that you are using the Islamic calendar. Your string is "1402/11/18" (reading in characters order). However, the library displays "18/11/1402" (so essentially performing unnecessary LTR -> RTL conversion). 

As far as I know, Arabic is the RTL language except for LTR numbers (therefore, the bidirectionality issue). Is there a similar rule for dates as well? Detecting numbers is simple, but dates may look like some part codes/invoice numbers, etc. How is it handled in other software? Maybe you need to inject a special Unicode character to hint at what to do? Maybe that should be done as part of culture conversion in dotnet?

https://www.w3.org/International/questions/qa-bidi-unicode-controls

**Comment by danirzv** [comment]
> @danirzv I am not an RTL expert, so please help me better understand this issue 😄 
> 
> I understand that you are using the Islamic calendar. Your string is "1402/11/18" (reading in characters order). However, the library displays "18/11/1402" (so essentially performing unnecessary LTR -> RTL conversion). 
> 
> As far as I know, Arabic is the RTL language except for LTR numbers (therefore, the bidirectionality issue). Is there a similar rule for dates as well? Detecting numbers is simple, but dates may look like some part codes/invoice numbers, etc. How is it handled in other software? Maybe you need to inject a special Unicode character to hint at what to do? Maybe that should be done as part of culture conversion in dotnet?
> 
> https://www.w3.org/International/questions/qa-bidi-unicode-controls

Actually its not doing rtl for that, input is as '18/11/1402' and should become rtl as '1402/11/18'

**Comment by navongol** [comment]
> 
Wrong,
Its not a standard approach for date in all rtl languages.
This issue is a date specific to arabic not to evry rtl language.
for example in hebrew a rtl language   18/11/1402 date  is correrct.

this date issue is specific language.

**Comment by MarcinZiabek** [comment]
@danirzv How are you formatting your date?

The following code:
```csharp
.Text(text =>
{
    text.Line(DateTime.Now.ToShortDateString());
    text.Line(DateTime.Now.ToString(CultureInfo.GetCultureInfo("pl").DateTimeFormat.ShortDatePattern));
    text.Span(DateTime.Now.ToString(CultureInfo.GetCultureInfo("ar").DateTimeFormat.ShortDatePattern));
});
```

Produces result that looks correct?

<img width="167" alt="image" src="https://github.com/QuestPDF/QuestPDF/assets/9263853/fab4ab68-1b11-4981-92f9-9354acaa08bc">

**Comment by MarcinZiabek** [comment]
This https://stackoverflow.com/a/67016414 comment mentions that slashes aren't regular characters and, therefore, alter bi-directionality.

**Comment by JimBobSquarePants** [comment]
How are you calculating the bidi level for each codepoint? Once you’ve done that you need to reorder the text based one those levels. You shouldn’t need to what type of character the codepoint represents. https://www.unicode.org/reports/tr9/#Reordering_Resolved_Levels

**Comment by MarcinZiabek** [comment]
@JimBobSquarePants In the 2024.3.0 release, I am removing the SkiaSharp dependency and directly using the Skia library. The text engine is powered by the SkParagraph module that uses HarfBuzz and ICU internally. I expect that text-related capabilities are now on par with Flutter, and likely highly accurate (or at least much better than it used to be).

**Comment by salawnah** [comment]
@MarcinZiabek Thank you :) I tested the alpha version with NopCommerce and it is working as expected

**Comment by hossein2408** [comment]
Please solve this problem

**Comment by AndreiMaz** [comment]
Is it fixed in the latest release?

**Comment by MarcinZiabek** [comment]
@AndreiMaz Yes, it has been fixed in the `2024.3.0` release. The bi-directional support is considerably better now 😄

**Comment by MarcinZiabek** [comment]
With the 2024.3 release, the text rendering engine has been completely redesigned, offering significantly improved functionality. It now supports nearly all common scenarios, including right-to-left languages, bidirectionality and text-shaping.

Since there has been no recent activity, I’m closing this issue. If you have a similar request in the future, feel free to open a new GitHub issue. Thanks for your feedback! ❤️
